File: Quick_Sort.java
import java.util.Arrays;
 public class Quick_Sort {
    public static void main(String[] args) {
        int []arr={1,22,34,4,55,42,8,-79};
        quicksort(arr,0, arr.length-1);
        System.out.println(Arrays.toString(arr));

    }
    public static void swap(int[]arr,int x,int y){
        int temp=arr[x];
        arr[x]=arr[y];
        arr[y]=temp;
    }
    //partition returns the correct index of the element and elements before that index are small and after that index are large
    static int partition(int[]arr,int st,int end){
        int pivot=arr[st];
        int count=0;
        for (int i = st+1; i <=end ; i++) {
            if(arr[i]<=pivot) count++;
        }
        int pivotInd=st+count;
        swap(arr,st,pivotInd);//yaha tak me pivot apni sahi place pr aagya h
        //aage ka code left side pr pivot se chote right side pr pivot se badhe elements daalega
        int i=st,j=end;//2 pointer
        while (i<pivotInd&&j>pivotInd){
            while (arr[i]<=pivot) i++;
            while (arr[j]>pivot) j--;
            if(i<pivotInd&&j>pivotInd){
                swap(arr,i,j);
                i++;
                j--;
            }
        }
        return pivotInd;
    }
    static void quicksort(int[]arr,int st,int end){
        if(st>=end){
            return;
        }
        int pi=partition(arr,st,end);
        quicksort(arr,st,pi-1);
        quicksort(arr,pi+1,end);
    } {
    
}
 }

File: C++/quicksort.cpp
#include <iostream>
using namespace std;

// Function to swap two elements
void swap(int* a, int* b) {
    int t = *a;
    *a = *b;
    *b = t;
}

// Function to partition the array
int partition(int arr[], int low, int high) {
    int pivot = arr[high];  // Choosing the last element as the pivot
    int i = (low - 1);      // Index of smaller element

    for (int j = low; j < high; j++) {
        // If current element is smaller than or equal to pivot
        if (arr[j] <= pivot) {
            i++;  // Increment index of smaller element
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);  // Place pivot in the correct position
    return (i + 1);
}

// QuickSort function
void quickSort(int arr[], int low, int high) {
    if (low < high) {
        // Partition the array
        int pi = partition(arr, low, high);

        // Recursively sort elements before and after partition
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

// Function to print the array
void printArray(int arr[], int size) {
    for (int i = 0; i < size; i++)
        cout << arr[i] << " ";
    cout << endl;
}

int main() {
    int arr[] = {10, 80, 30, 90, 40, 50, 70};
    int n = sizeof(arr) / sizeof(arr[0]);

    cout << "Original array: ";
    printArray(arr, n);

    quickSort(arr, 0, n - 1);

    cout << "Sorted array: ";
    printArray(arr, n);

    return 0;
}
